src/dune_engine/scheduler.ml

```diff
@@ -640,8 +664,98 @@ module Handler = struct
   type t = Config.t -> Event.t -> unit
 end
 
+module Alarm_clock : sig
+  type t
+
+  val create : Event.Queue.t -> frequency:float -> t
+
+  type alarm
+
+  val await : alarm -> [ `Finished | `Cancelled ] Fiber.t
+
+  val cancel : t -> alarm -> unit
+
+  val sleep : t -> float -> alarm
+
+  val close : t -> unit
+end = struct
+  type alarm = [ `Finished | `Cancelled ] Fiber.Ivar.t
+
+  type t =
+    { events : Event.Queue.t
+    ; mutex : Mutex.t
+    ; frequency : float
+    ; mutable alarms : (float * [ `Finished | `Cancelled ] Fiber.Ivar.t) list
+    ; mutable active : bool
+    }
+
+  let await = Fiber.Ivar.read
+
+  let cancel t alarm =
+    Mutex.lock t.mutex;
+    let found = ref false in
+    t.alarms <-
+      List.filter t.alarms ~f:(fun (_, alarm') ->
+          let eq = alarm' == alarm in
+          if eq then found := true;
+          not eq);
+    Mutex.unlock t.mutex;
+    if !found then
+      Event.Queue.send_timers_completed t.events
+        [ Fiber.Fill (alarm, `Cancelled) ]
+
+  let polling_loop t () =
+    let rec loop () =
+      match t.active with
+      | false -> ()
+      | true ->
+        let now = Unix.gettimeofday () in
+        let expired, active =
+          List.partition_map t.alarms ~f:(fun (expiration, ivar) ->
+              if now < expiration then
+                Left (Fiber.Fill (ivar, `Finished))
+              else
+                Right (expiration, ivar))
+        in
+        t.alarms <- active;
+        Mutex.unlock t.mutex;
+        if active <> [] then Event.Queue.send_timers_completed t.events expired;
```

Comment:
Meant to check `expired <> []` here? That's probably why the CI times out.
By the way, do you know if we can reduce the timeout from 6 hours to, say, one hour?

Reply:
> Also, you can use a simple pattern-match here. Why use polymorphic equality when you can easily avoid it?

The pattern match is a couple of lines longer :)

> you could make the type accepted by send_timers_completed a Non_empty_list.t and then you can't make this error.

Good idea. I keep forgetting that we introduced this.